[py3] Fix LDIF parsing

rbarrois opened the pull request:
Follow the LDIF RFC (https://tools.ietf.org/html/rfc2849):

- A LDIF file MUST contain only safe ASCII
- Any known text value MUST be UTF-8 + base64 encoded
- Any other value (bytes) MUST be base64 encoded

Notes:

- We had to change a test, which was trying to parse an invalid LDIF
  file containing non-ascii text
- The LDIFParser will return:
  * Text for operations, attribute names, distinguisedNames, controls
  * Bytes for all attribute values

rbarrois committed:
[py3] Fix LDIF parsing

Follow the LDIF RFC (https://tools.ietf.org/html/rfc2849):

- A LDIF file MUST contain only safe ASCII
- Any known text value MUST be UTF-8 + base64 encoded
- Any other value (bytes) MUST be base64 encoded

Notes:

- We had to change a test, which was trying to parse an invalid LDIF
  file containing non-ascii text
- The LDIFParser will return:
  * Text for operations, attribute names, distinguisedNames, controls
  * Bytes for all attribute values

encukou commented:
I believe it makes sense to allow _reading_ UTF-8 encoded LDIF files as a _non-standard extension_ inherited from python-ldap. UTF-8 is pretty widespread, a proper superset of ASCII, and hard to confuse for other encodings. I believe following the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) would be a good thing to do here.

rbarrois commented:
To be sure, you mean "allow reading a LDIF files containing UTF-8 encoded values"?

The issue is that this could lead to silent failures with files in other encodings, as there would be no way to  find out whether the file contains UTF-8, ISO-8859-1 or other ;)

I suggest we raise the question on the python-ldap mailing list, and adjust accordingly; what do you think?

encukou commented:
Hm, maybe I misunderstood the problem. I guess the behavior should be:

The user can pass a file opened in text mode (with any encoding they want); then pyldap gets text and doesn't need to worry about encoding – it will only decode ASCII-only values as needed. Or they can pass a file opened in bytes mode; then it can be ASCII-only.

Do you agree with that as the indended behavior?

rbarrois commented:
Yes, I agree with this ;)

What I mean is that the following file would not be a valid _LDIF_ file, because it contains non-ascii bytes (the `ö`), which is forbidden by the standard — it should be base64-encoded:

``` ini
dn: cn=Michael Stroeder,dc=stroeder,dc=com
lastname: Ströder
```

On the other hand, the following form is valid:

``` ini
dn: cn=Michael Stroeder,dc=stroeder,dc=com
lastname:: U3Ryw7ZkZXI=
```

The changes I propose mean that:
- A _valid_ LDIF file can be handled, no matter whether it was opened as `r` or `rb`
- An _invalid_ LDIF file (i.e containing non-ASCII bytes), will be rejected, whichever mode it was opened in

encukou commented:
OK, that's where we disagree. I think a text file with non-ASCII values should be handled successfuly. In that case there should be no confusion as to what encoding it is. There are files like that in the wild, and removing the ability to process them would be one more hurdle to switching from python-ldap to pyldap (and so from py2 to py3).

rbarrois commented:
OK, seems fair.

I'll update the PR accordingly.

rbarrois committed:
Restore support for UTF-8 values in ldif.

Even if this is not supported in the standards, it is a common
exception, and is supported by our upstream.

encukou commented:
Thanks a lot!

I've pushed to master. Can you think of anything else to fix before releasing `2.4.27`?
